### PR TITLE
type symbols now use ID_symbol_type

### PR DESCRIPTION
**Pull request by kroening**
This changes the ID of symbol_typet from ID_symbol to ID_symbol_type to avoid the clash with symbol_exprt().

It should be considered to rename it all to typedef_type, which is a better description of what this type does.

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]
Changes are as described and do not appear harmful.

**Comment by kroening** [comment]
This is now ready to go.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 2b81e4e7e87ae9690ec79331298a3ac7f675e2e7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79644083

File: src/analyses/uncaught_exceptions_analysis.cpp [hunk 1]
```diff
@@ -19,11 +19,10 @@ irep_idt uncaught_exceptions_domaint::get_exception_type(const typet &type)
 {
   PRECONDITION(type.id()==ID_pointer);
 
-  if(type.subtype().id()==ID_symbol)
-  {
+  if(type.subtype().id() == ID_symbol_type)
     return to_symbol_type(type.subtype()).get_identifier();
-  }
-  return ID_empty;
+  else
```
Comment (thk123) [inline src/analyses/uncaught_exceptions_analysis.cpp:24]: Nit: Unrelated change should either not be done or go in its own commit

File: src/ansi-c/ansi_c_declaration.cpp [hunk 2]
```diff
@@ -22,12 +22,12 @@ void ansi_c_declaratort::build(irept &src)
...
   typet *p=static_cast<typet *>(&src);
 
-  // walk down subtype until we hit symbol or "abstract"
+  // walk down subtype until we hit symbol_type or "abstract"
   while(true)
   {
     typet &t=*p;
 
-    if(t.id()==ID_symbol)
+    if(t.id() == ID_symbol_type || t.id() == ID_symbol)
```
Comment (thk123) [inline src/ansi-c/ansi_c_declaration.cpp]: Why do we allow both `symbol_type` and `symbol` here but no where else?

**Comment by smowton** [comment]
Disagree with typedef-type because at least in the Java frontend there is no such operation, but we still use it as a type abbreviation -- also Ada uses this lots, etc. `type_symbol` is good IMO.

This will definitely break test-gen. @thk123 suggest making a real bump.

**Review by thk123 — DISMISSED** [review]
This is definitely going to need a real TG bump to presumably apply the same change there.

File: src/ansi-c/ansi_c_declaration.cpp (hunk 2 above)
Comment (thk123) [reply src/ansi-c/ansi_c_declaration.cpp]: Having now reviewed the whole PR I believe this is a mistake

File: jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp [hunk 3]
```diff
@@ -112,7 +112,7 @@ void ci_lazy_methods_neededt::gather_field_types(
   {
     // If class_type is not a symbol this may be a reference array,
     // but we can't tell what type.
-    if(class_type.id() == ID_symbol)
+    if(class_type.id() == ID_symbol_type)
```
Comment (thk123) [inline jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp:115]: To avoid this pain in the future, in the TG bump I am going to use `can_cast_type<symbol_typet>(class_type)` rather than directly accessing the ID - you might like to do something similar with this PR while the TG bump is being prepared :blush:

Comment (kroening) [reply jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp:115]: This should be one-off -- we should simply stick to the convention that the id() matches the name of the class type.

File: src/ansi-c/ansi_c_declaration.cpp (hunk 2 above)
Comment (kroening) [reply src/ansi-c/ansi_c_declaration.cpp]: There is indeed ambiguity in the front-end whether an identifier is a typedef or not. The standard example is

```
typedef int some_identifier;

int main()
{
  int some_identifier;

}
```

**Comment by kroening** [comment]
@smowton  Indeed, there are no typedefs in Java byte code. The idea is to use struct_tag (and union_tag) instead of the symbol_type for pointers to structs/unions/classes. The C and C++ front-ends would use typedef_type for proper typedefs.

File: jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp (hunk 3 above)
Comment (thk123) [reply jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp:115]: > should be considered to rename it all to typedef_type

Seems like it could change at least one more time. 

In any case, this pattern, were it to be used more widely, would hide more of the internal ireps and make changes like this less painful in future. 

It also allows for at a later date adding more rigorous type checking (e.g. a `symbol_typet` should always have one named sub, `ID_identifier` etc).

File: src/ansi-c/ansi_c_declaration.cpp (hunk 2 above)
Comment (thk123) [reply src/ansi-c/ansi_c_declaration.cpp]: Sorry still don't get it - how can a `typet` ever have an `id` of `ID_symbol`? I thought that was now reserved exclusively for `symbol_exprt`s?

Comment (kroening) [reply src/ansi-c/ansi_c_declaration.cpp]: ansi_c_declaratort::build runs well before the class invariants for typet are established -- I'll change this to irept to make this clearer.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 2855c2abc31395b5eed50fa415c57598913ae1a5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79831442
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 3a9ca834f40ac1d55cabf738c5946865f5f1fbfd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79834736
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Comment by thk123** [comment]
TG bump: diffblue/test-gen#2086 not yet passed

**Comment by thk123** [comment]
Update: TG bump has not passed and I don't really know why, first thing to try is a rebase, could you rebase this?

**Comment by kroening** [comment]
rebase done

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 70d8d6727b4953d145e430ada7b95bc82840c0eb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80101153

**Comment by thk123** [comment]
TG bump updated - hopefully it will now pass.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 4]
```diff
@@ -1635,7 +1635,7 @@ codet java_bytecode_convert_methodt::convert_instructions(
       const typecast_exprt pointer(op[0], java_array_type(statement[0]));
 
       dereference_exprt array(pointer, pointer.type().subtype());
-      assert(pointer.type().subtype().id()==ID_symbol);
+      assert(pointer.type().subtype().id() == ID_symbol_type);
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: `PRECONDITION`

File: src/ansi-c/c_typecheck_initializer.cpp [hunk 5]
```diff
@@ -492,7 +492,7 @@ exprt::operandst::const_iterator c_typecheck_baset::do_designated_initializer(
 
     const typet &type=designator.back().subtype;
     const typet &full_type=follow(type);
-    assert(full_type.id()!=ID_symbol);
+    assert(full_type.id() != ID_symbol_type);
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_initializer.cpp]: `CHECK_RETURN`

File: src/goto-instrument/goto_program2code.cpp [hunk 6]
```diff
@@ -1861,7 +1861,7 @@ void goto_program2codet::cleanup_expr(exprt &expr, bool no_typecast)
     if(no_typecast)
       return;
 
-    assert(expr.type().id()==ID_symbol);
+    assert(expr.type().id() == ID_symbol_type);
```
Comment (tautschnig) [inline src/goto-instrument/goto_program2code.cpp]: `INVARIANT`

**Review by tautschnig — CHANGES_REQUESTED** [review]
This will break compatibility with previously compiled goto binaries. I would suggest that we increase the goto binary version with this change (as in general we should be more aggressive with bumping the version number) to avoid surprises.

**Comment by kroening** [comment]
Both done

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
Still needs some fixes as CI currently fails, but otherwise ok from my point of view. Once that's cleared the TG bump needs to be checked.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 449b04839a409be5b87ff173a746b37ede0efa2d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80482269

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d7b279cf08e9a344529e761b03e3275fb475e58f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80488850

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 28dee3581d5499117e05789ea89e23a2a8dbe23c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80868219

**Comment by LAJW** [comment]
This passes on TG now (we've found all instances of `ID_symbol` that have to be replaced).

**Review by thk123 — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]
